ee-9/source-transform/ee/src/main/java/org/jboss/as/ee/concurrent/deployers/EEConcurrentContextProcessor.java

```diff
@@ -103,7 +107,15 @@ private void setupConcurrentContext(ConcurrentContext concurrentContext, String
             concurrentContext.addFactory(factory);
         }
         concurrentContext.addFactory(new OtherEESetupActionsContextHandleFactory(deploymentUnit.getAttachmentList(Attachments.OTHER_EE_SETUP_ACTIONS)));
-
+        // add factories for deployment provided thread context providers
+        final ServiceLoader<ThreadContextProvider> threadContextProviderServiceLoader = ServiceLoader.load(ThreadContextProvider.class, moduleClassLoader);
+        final Iterator<ThreadContextProvider> threadContextProviderIterator = threadContextProviderServiceLoader.iterator();
+        int count = 0;
+        while (threadContextProviderIterator.hasNext()) {
+            final ThreadContextProvider threadContextProvider = threadContextProviderIterator.next();
+            concurrentContext.addFactory(new ThreadContextProviderContextHandleFactory(threadContextProvider, count));
+            count++;
+        }
```

Comment:
This iteration should likely be wrapped in a privileged action if the security manager is present. It looks like it uses the modules class loader which could be a deployment.